internal/services/apimanagement/api_management_policy_fragment_resource.go

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package apimanagement
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"time"
+
+	"github.com/hashicorp/go-azure-helpers/lang/pointer"
+	"github.com/hashicorp/go-azure-helpers/lang/response"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonschema"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/apimanagement/2022-08-01/apimanagementservice"
+	"github.com/hashicorp/go-azure-sdk/resource-manager/apimanagement/2022-08-01/policyfragment"
+	"github.com/hashicorp/terraform-provider-azurerm/helpers/tf"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/services/apimanagement/schemaz"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/pluginsdk"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tf/validation"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/timeouts"
+)
+
+func resourceApiManagementPolicyFragment() *pluginsdk.Resource {
+	return &pluginsdk.Resource{
+		Create: resourceApiManagementPolicyFragmentCreate,
+		Read:   resourceApiManagementPolicyFragmentRead,
+		Update: resourceApiManagementPolicyFragmentUpdate,
+		Delete: resourceApiManagementPolicyFragmentDelete,
+		Importer: pluginsdk.ImporterValidatingResourceIdThen(func(id string) error {
```

Comment:
thanks for your review! 
I think I implemented all your remarks with the exception of the import and format topic and I left one comment to on of your remarks as well.

I'll take another look at the import and format topic and provide my results.

Reply:
@stephybun can you help look into the changes that was put in a week ago upon your request?

Reply:
@travisgan as far as I can tell the import function hasn't been updated yet and the tests would still fail on the import check. 

@jan-mrm have you had a chance to look into this yet?

Reply:
@stephybun I think I am stuck at this point. 🫤😮‍💨

The overarching problem is the format and how to handle it.

> Format is supplied as an option, but is also set and I assumed returned in the request and response bodies of the Properties model, so wouldn't we be able to deduce the format from that field?

I don't think so. As far as I tested it, the apim-service stores all policy fragments in ONE format (could be rawxml or xml). It always displays the fragment interpreted as rawxml (so lets assumed its saved that way also). It returns the fragment in the format defined in the GET request. So if you set the format in a GET request to xml (or omit it since its the default) you get the stored fragment in xml format. If you ask for the rawxml version you get it as rawxml. (The fragment does not know anymore, what format it was created as, if that makes sense.)

In the following table I omit the wrapping \<fragment>\</fragment>.
Also the " is escaped in the actual https request by \".
Using the [fragment](https://stackoverflow.com/questions/70278751/what-is-the-difference-between-xml-and-rawxml-formats-when-defining-apim-pol/70310169#70310169) as an example

| nr | fragment value | PUT format | displayed in azure portal | GET as xml | GET as rawxml |
| --- | --- | --- | --- | --- | --- |
| 1 | \<set-variable name="var" value="@(\&quot;user id:\&quot; + context.User?.Id)" /> | xml | \<set-variable name="var" value="@("user id:" + context.User?.Id)" /> | \<set-variable name="var" value="@(\&quot;user id:\&quot; + context.User?.Id)" /> | \<set-variable name="var" value="@("user id:" + context.User?.Id)" /> |
| 2 | \<set-variable name="var" value="@("user id:" + context.User?.Id)" /> | rawxml | <set-variable name="var" value="@("user id:" + context.User?.Id)" /> | \<set-variable name="var" value="@(\&quot;user id:\&quot; + context.User?.Id)" /> | <set-variable name="var" value="@("user id:" + context.User?.Id)" /> |
| 3 | \<set-variable name="var" value="@(\&quot;user id:\&quot; + context.User?.Id)" /> | rawxml | \<set-variable name="var" value="@(\&quot;user id:\&quot; + context.User?.Id)" /> | \<set-variable name="var" value="@(\&amp;quot;user id:\&amp;quot; + context.User?.Id)" /> | \<set-variable name="var" value="@(\&quot;user id:\&quot; + context.User?.Id)" /> |
| 4 | \<set-variable name="var" value="@("user id:" + context.User?.Id)" /> | xml | - | - | - |

nr 1 and 2 result in the same.
In nr 3 the & is further escaped.
nr 4 is accepted with a 202 by the api (like all the the other fragments) but never actually created. Since its not actually xml format.

---

I did set `ForceNew: true` for the `format` but this seems like a bad idea, since the API allows changing the format and value. I just did that to avoid weird states, but id want to revert that if possible.

So when removing that I get weird effects when changing the format and/or the value of a fragment.

Which in the end comes down to the `DiffSuppressFunc: XmlWithDotNetInterpolationsDiffSuppress,` which potentially suppresses too much...

I can take another look into this.

---

The import problem might resolve itself, when not forcing a recreate and updating the diff. But if not I'm not sure how to know which format to use when someone imports a fragment. I feel like it would have to be the users input.

---

But help is appreciated (or someone having another go at adding the resource, maybe with a fresh view) 🙂
maybe I'm also looking at it wrong

Reply:
Thanks for all the details @jan-mrm!

Since we're unable to distinguish whether a policy fragment was created as `xml` or `rawxml` at import time I think the current implementation is the best that we can do for now, i.e. always importing as the default format `xml`. As a result I agree that `format` should no longer be `ForceNew`, that way users would at least be able to update the imported fragment to the format that it should be, we should add a note to the docs on this. 

Due to this behaviour there are some additional changes needed in the test, also having looked through this again there are some more consistency improvements that should be made to the resource. I will leave those as comments in-line.

Reply:
guessing this can be closed then